test/sandbox-test.js

```diff
@@ -2377,4 +2377,52 @@ describe("Sandbox", function () {
             sandboxB.restore();
         });
     });
+    describe("concurrency safe", function () {
+        it("spy", async function () {
+            class F {
+                constructor() {
+                    this.first = [];
+                    this.second = [];
+                    this.third = [];
+                }
+                async run() {
+                    await this.execute("first");
+                    await Promise.resolve();
+                    await this.execute("second");
+                    await Promise.resolve();
+                    await this.execute("third");
+                }
+
+                async execute(s) {
+                    for (const f of this[s]) {
+                        await f();
+                    }
+                }
```

Comment:
This test was a bit hard to wrap my head around, at least at midnight ... 😄 Could you clarify what you are really doing? I mean, I assume you are somehow testing quazi-concurrent runs, interlacing runs from different sandboxes, but I still find it a bit hard to understand and read.

Is there some way you could re-order or refactor this in some way that made the intent as expressed through the code clearer? 

To me, it is not at all clear how the sandbox runs are being interlaced and in what order functions are being called. Would it not be much clearer if doing something a la this pseudo-code

```js
const sandboxASpy1 = sandboxA.spy();
const sandboxASpy2 = sandboxA.spy();
const sandboxBSpy1 = sandboxB.spy();
const sandboxBSpy2 = sandboxB.spy();

sandboxASpy1();
sandboxBSpy1();
sandboxBSpy2();
sandboxASpy2();

assert(sandboxASpy1.calledImmediatelyBefore(sandboxASpy2));
assert(sandboxBSpy1.calledImmediatelyBefore(sandboxBSpy2));
```

There is no need for awaits AFAIK. JS is single threaded no matter what, and the asynchronicity just muddles up what is going on IMHO.